pkg/controller/storage/capability/csi_capability.go

```diff
@@ -0,0 +1,166 @@
+	pluginCapabilitiesResponse, err := identityClient.GetPluginCapabilities(ctx, &csi.GetPluginCapabilitiesRequest{})
+	if err != nil {
+		return err
+	}
+
+	for _, capability := range pluginCapabilitiesResponse.GetCapabilities() {
+		if capability == nil {
+			continue
+		}
+		if capability.GetService().GetType() == csi.PluginCapability_Service_VOLUME_ACCESSIBILITY_CONSTRAINTS {
+			spec.Features.Topology = true
+		}
+		volumeExpansion := capability.GetVolumeExpansion()
+		if volumeExpansion != nil {
+			switch volumeExpansion.GetType() {
+			case csi.PluginCapability_VolumeExpansion_ONLINE:
+				spec.Features.Volume.Expand = v1alpha1.ExpandModeOnline
+			case csi.PluginCapability_VolumeExpansion_OFFLINE:
+				spec.Features.Volume.Expand = v1alpha1.ExpandModeOffline
+			}
+		}
+	}
+	return nil
+}
+
+func addControllerCapabilities(ctx context.Context, conn *grpc.ClientConn, spec *v1alpha1.StorageClassCapabilitySpec) error {
+	controllerClient := csi.NewControllerClient(conn)
+	controllerCapabilitiesResponse, err := controllerClient.ControllerGetCapabilities(ctx, &csi.ControllerGetCapabilitiesRequest{})
+	if err != nil {
+		return err
+	}
+	for _, capability := range controllerCapabilitiesResponse.GetCapabilities() {
+		switch capability.GetRpc().GetType() {
+		case csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME:
+			spec.Features.Volume.Create = true
+		case csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME:
+			spec.Features.Volume.Attach = true
+		case csi.ControllerServiceCapability_RPC_LIST_VOLUMES:
+			spec.Features.Volume.List = true
+		case csi.ControllerServiceCapability_RPC_CLONE_VOLUME:
+			spec.Features.Volume.Clone = true
+		case csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT:
+			spec.Features.Snapshot.Create = true
+		case csi.ControllerServiceCapability_RPC_LIST_SNAPSHOTS:
+			spec.Features.Snapshot.List = true
+		}
+	}
+	return nil
+}
+
+func addNodeCapabilities(ctx context.Context, conn *grpc.ClientConn, spec *v1alpha1.StorageClassCapabilitySpec) error {
+	nodeClient := csi.NewNodeClient(conn)
+	controllerCapabilitiesResponse, err := nodeClient.NodeGetCapabilities(ctx, &csi.NodeGetCapabilitiesRequest{})
+	if err != nil {
+		return err
+	}
+	for _, capability := range controllerCapabilitiesResponse.GetCapabilities() {
+		switch capability.GetRpc().GetType() {
+		case csi.NodeServiceCapability_RPC_GET_VOLUME_STATS:
+			spec.Features.Volume.Stats = true
+		}
+	}
+	return nil
+}
+
+// Connect address by GRPC
+func connect(address string) (*grpc.ClientConn, error) {
+	dialOptions := []grpc.DialOption{
+		grpc.WithInsecure(),
+	}
+	u, err := url.Parse(address)
+	if err == nil && (!u.IsAbs() || u.Scheme == "unix") {
+		dialOptions = append(dialOptions,
+			grpc.WithDialer(
+				func(addr string, timeout time.Duration) (net.Conn, error) {
+					return net.DialTimeout("unix", u.Path, timeout)
+				}))
+	}
+	// This is necessary when connecting via TCP and does not hurt
```

Comment:
what does `hurt` mean here? Compatible?

Reply:
I use the func from [csi-test](https://github.com/kubernetes-csi/csi-test/blob/master/utils/grpcutil.go#L31). I think 'does not hurt' means 'the dialOption has no influence' when using Unix domain socket. If this is not suitable，I will delete it.